src/node_order_optim/node_permute.cpp

```diff
@@ -489,4 +474,5 @@ void permute_nodeindices(Memb_list* ml, int* p) {
     permute_ptr(ml->nodeindices, ml->nodecount, ml->_permute);
```

Comment:
> though I think the size always refers to the data and not the perm vector. So perhaps size should be the second arg. But I'm not sure that would make anything more clear.

Good point :+1: Maybe the following makes more sense:

```
void forward_permute(T*& data, , int data_size, const std::vector<int>& perm)
```

Btw reading a comment from this file I got a bit confused. In https://github.com/neuronsimulator/nrn/blob/39c8cce0429bd4615b0e9bc8a6138478c6be69e8/src/node_order_optim/node_permute.cpp#L33 it says:

> Notice that p on the left side is a forward permutation. On the right side
it serves as the inverse permutation.
area_original[i] <- area_permuted[p[i]]

I thought it was the other way around? I created a unit test to try and clarify this.

> Is it too tedious at this time to modify NrnThread to only use std::vector<...> for it's arrays?

I will have a look tomorrow to evaluate how much work this would need. If it's not a lot of work it would make sense to come together or before the permutation PR

Reply:
It is the other way around! When I first wrote the code I reversed the ?normal? convention. Ollie noticed that and mentioned it in the comment in node_permute.cpp
```
/*
Below, the sense of permutation, is reversed. Though consistent, forward
permutation should be defined as (and the code should eventually transformed)
so that
  v: original vector
  p: forward permutation
  pv: permuted vector
  pv[i] = v[p[i]]
and
  pinv: inverse permutation
  pv[pinv[i]] = v[i]
Note: pinv[p[i]] = i = p[pinv[i]]
*/
```
Every once in a while I try to get my intuition to accept that a forward  permutation is ```new_a[i] = old_a[p[i]]```
For example, imagine an old vector  ```[a, b, c, d]```. Then I would think a rotation one step the right would be a forward permutation and the new vector would be ```[d, a, b, c]```. i.e. a moves from 0 to 1. i.e ```new[1] = old[0]```, etc. So the p vector is
```[3, 0, 1, 2]```. So the permutation vector is clearly rotation of one step to the right of the identity permutation ```[0,1,2,3]```.  To avoid confusion, I think it is really important to repair all my old convention reversal mistakes (and fix up comments as well).   At that point it will be interesting to see if ONLY forward permutations are used and there is no need for any reverse permutation implementations.

Reply:
It's true that the `inverse_permute` I implemented is not used anywhere at the moment since it didn't exist before. However there are multiple places were we invert the permute vector and we then do the `forward_permute`. In those cases we can use instead the `inverse_permute` function.
For example: https://github.com/neuronsimulator/nrn/blob/dccc4565e74e0819922e8478de94a375f78ea477/src/node_order_optim/node_permute.cpp#L473
I also noticed that you replaced the call to `permute_nodeindices()` in https://github.com/neuronsimulator/nrn/blob/dccc4565e74e0819922e8478de94a375f78ea477/src/coreneuron/io/phase2.cpp#L1161 (this is the CoreNEURON code) with https://github.com/neuronsimulator/nrn/blob/dccc4565e74e0819922e8478de94a375f78ea477/src/node_order_optim/cellorder.cpp#L432 (in the NEURON code). I still can't comprehend how these two are identical or they shouldn't be?

Reply:
In coreneuron/io/phase2.cpp there are two for loops over all the tml. The first
```
                permute_nodeindices(tml->ml, p)
```
calls ```node_permute``` to do the inplace ```vec[i] = permute[vec[i]];```
then sorts the nodeindices in increasing order and stores that as the ml's permutation vector
```ml->_permute = nrn_index_sort(ml->nodeindices, ml->nodecount);```
(```_permute``` is then inverted and used in the call to ``` permute_ptr``` to permute the nodeindices (using the confusing ```void permute(T* data, int cnt, int sz, int layout, int* p) {```  which uses the (old wrong?) convention ```data[ip] = data_orig[i]```

The second loop calls ```permute_ml``` to permute the ml data and pdata

In node_order_optim/cellorder.cpp the equivalent of the first loop is to do the in-place part of the update with
```
ml->nodeindices[i] = perm[ml->nodeindices[i]];
```
and finish off by sorting the nodeindices and the other ml arrays (nodelist, prop, and pdata)  by calling ```sort_ml(ml);```

The tests I've been using is the temporary nrn/test.py that I examine with 1,2,3 for pc.optimize_node_order(1)
and also verify that nrntraub is generating identical spike trains.